### PR TITLE
feat: Add webhook-xblock

### PR DESCRIPTION
**Pull request by magajh**
This XBlock triggers a webhook by sending a payload with basic information about the course and student to a configured URL.

The XBlock can be set in three modes:
- Send the payload every time the user visits the course unit
- Send the payload only the first time the user visits the course unit (default mode)
- Allow the student to send the request by clicking on a button

The view from Studio allows the staff user to configure multiple things
studio1
studio2
studio3

In case the XBlock is configured to allow the student to send the request by clicking a button, this is how the course unit would look to the student
studio4
Course unit (student's view)
button

**UPDATE**: I am now sending a flat dictionary in the payload since it may be difficult for some webhooks to handle nested dictionaries

Example of the request data sent in the payload
```
{
  'student_username': 'admin', 
  'student_date_joined': '2021-07-22T08:40:07-05:00', 
  'timestamp': '2021-07-22T08:51:53.725556', 
  'student_is_active': True, 
  'student_email': 'time@example.com', 
  'anonymous_student_id': 'a882a4d4a7bb574b3ced83a63a9f7f97', 
  'course_id': 'course-v1:edunext+01+test'
}
```

**UPDATE**: I hid the *send async* option since celery is not registering the tasks from the xblock. I will uncomment and add the field again once this issue gets solved.

### CODE REVIEW

File: webhook_xblock/static/js/src/webhook_xblock.js [hunk 1]
```diff
@@ -1,28 +1,45 @@
...
+            button.type = 'button';
+            button.innerHTML = button_text;
+            button.className = 'btn-styled';
+            
+            button.onclick = function() {
+                $.ajax({
+                    type: "POST",
+                    url: handlerUrl,
+                    data: JSON.stringify({'send_payload': 'True'}),
+                    success: notifyStudent
```
Comment (mariajgrimaldi) [inline webhook_xblock/static/js/src/webhook_xblock.js:30]: Even when the student does not send it?

Comment (magajh) [reply webhook_xblock/static/js/src/webhook_xblock.js:30]: @mariajgrimaldi I think I'm going to disable this button when running from studio mode and show the user (staff) a proper message

File: webhook_xblock/static/html/webhook_xblock_edit.html [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    <li class="field comp-setting-entry is-set">
+        <div class="wrapper-comp-setting">
+          <label class="label setting-label" for="send_course_grade">Send Course Grade</label>
+          <input class="input setting-input" style="margin-left: -270pt;" name="send_course_grade" id="send_course_grade" value="{send_course_grade}" type="checkbox" />
+        </div>
+        <span class="tip setting-help">Send the student's course grade in the payload.</span>
+    </li>
+    <li class="field comp-setting-entry is-set">
+      <div class="wrapper-comp-setting">
+        <label class="label setting-label" for="send_async">Send in Async Task</label>
```
Comment (mariajgrimaldi) [inline webhook_xblock/static/html/webhook_xblock_edit.html]: where can we see this? it's not in the screenshot

File: webhook_xblock/webhook_xblock.py [hunk 3]
```diff
@@ -1,41 +1,90 @@
...
         """
         html = self.resource_string("static/html/webhook_xblock.html")
-        frag = Fragment(html.format(self=self))
+        frag = Fragment(html.format(
+            self=self, 
+            frequency=self.frequency, 
+            button_text=self.button_text,
+            component_text=self.component_text,
+        ))
         frag.add_css(self.resource_string("static/css/webhook_xblock.css"))
```
Comment (mariajgrimaldi) [inline webhook_xblock/webhook_xblock.py:178]: do you have this file?

Comment (magajh) [reply webhook_xblock/webhook_xblock.py:178]: @mariajgrimaldi yes, it's the same from the cookie-cutter, so you can see it in the source code (master)

File: webhook_xblock/static/html/webhook_xblock_edit.html (hunk 2 above)
Comment (magajh) [reply webhook_xblock/static/html/webhook_xblock_edit.html]: @mariajgrimaldi you are right, I added the rest of the screenshots so you can see all the configurations available from the studio view

**Review by MoisesGSalas — COMMENTED** [review]
Can you add a title to the xblock? so it doesn't show that long string. IIRC you have to add a  new field `display_name`.

**Comment by mariajgrimaldi** [comment]
do you have a real example of extra information?

File: webhook_xblock/static/html/webhook_xblock_edit.html [hunk 4]
```diff
@@ -0,0 +1,69 @@
+<div class="wrapper-comp-settings is-active editor-with-buttons" id="settings-tab">
+  <ul class="list-input settings-list">
+    <li class="field comp-setting-entry is-set">
+      <div class="wrapper-comp-setting">
+        <label class="label setting-label" for="webhook_url">URL</label>
+        <input class="input setting-input" style="width: 550px;" name="webhook_url" id="webhook_url" value="{webhook_url}" type="url" required/>
```
Comment (mariajgrimaldi) [inline webhook_xblock/static/html/webhook_xblock_edit.html]: Why don't you use the .css file instead of inline style?

**Comment by mariajgrimaldi** [comment]
can you also update the readme.rst?

File: webhook_xblock/static/js/src/webhook_xblock.js (hunk 1 above)
Comment (magajh) [reply webhook_xblock/static/js/src/webhook_xblock.js:30]: @mariajgrimaldi this was fixed

File: webhook_xblock/static/html/webhook_xblock_edit.html (hunk 4 above)
Comment (magajh) [reply webhook_xblock/static/html/webhook_xblock_edit.html]: @mariajgrimaldi done!

File: webhook_xblock/utils.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import collections
+
+
+def flatten_dict(d, parent_key='', sep='_'):
+    """
+    This function returns a flatten dictionary-like object.
+    """
+    items = []
+    for key, value in d.items():
+        new_key = parent_key + sep + key if parent_key else key
```
Comment (MoisesGSalas) [inline webhook_xblock/utils.py]: This is a string I guess. Try to use `format()`

**Comment by magajh** [comment]
@mariajgrimaldi right now I don't have a real example of extra information, maybe it could be useful for sending some info related to the course. Do you think it's unnecessary?

File: webhook_xblock/utils.py (hunk 5 above)
Comment (magajh) [reply webhook_xblock/utils.py]: @MoisesGSalas  Done

Comment (mariajgrimaldi) [reply webhook_xblock/utils.py]: Did you use `format`?

**Comment by mariajgrimaldi** [comment]
> @mariajgrimaldi right now I don't have a real example of extra information, maybe it could be useful for sending some info related to the course. Do you think it's unnecessary?

Not at all :100:

File: webhook_xblock/static/css/webhook_xblock.css [hunk 6]
```diff
@@ -1,9 +1,14 @@
...
 
-.webhook_xblock_block p {
-    cursor: pointer;
+.input setting-input-bool {
+    margin-left: -270pt;
 }
+
+.input setting-input {
+    width: 550px;
+}
```
Comment (andrey-canon) [inline webhook_xblock/static/css/webhook_xblock.css]: :( missing line

File: webhook_xblock/static/js/src/webhook_xblock.js [hunk 7]
```diff
@@ -1,28 +1,51 @@
 /* Javascript for WebhookXblock. */
 function WebhookXblock(runtime, element) {
+    var frequency = document.getElementById('frequency').value;
```
Comment (andrey-canon) [inline webhook_xblock/static/js/src/webhook_xblock.js]: could you add a prefix to the ids something like  `webhook-xblock-frecuency`

**Review by mariajgrimaldi — APPROVED** [review]

File: webhook_xblock/tasks.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+from celery.utils.log import get_task_logger
+
+logger = get_task_logger(__name__)
+
+RETRY_DELAY = 5 * 60
+MAX_RETRIES = 3
+REQUEST_TIMEOUT = 10
+
+
+@task(max_retries=MAX_RETRIES)
```
Comment (andrey-canon) [inline webhook_xblock/tasks.py]: I think this should be shared_task, also you don't nee to handle the exception if you include the parameter on the decorator https://docs.celeryproject.org/en/stable/userguide/tasks.html#automatic-retry-for-known-exceptions  please use exponential_backoff

**Review by MoisesGSalas — APPROVED** [review]
LGTM. But try too run a linter

File: webhook_xblock/webhook_xblock.py [hunk 9]
```diff
@@ -1,41 +1,93 @@
...
+
+        return False
+
+    def make_request(self, data):
+        """
+        Helper for sending the request to the payload.
+        """
+        response = requests.post(self.webhook_url, data)
+
+        return response.status_code == 200
```
Comment (andrey-canon) [inline webhook_xblock/webhook_xblock.py]:
```suggestion
        return response.ok
```

File: webhook_xblock/webhook_xblock.py [hunk 10]
```diff
@@ -47,21 +99,73 @@ def student_view(self, context=None):
...
+            payload.update(self.extra_info)
+            payload = flatten_dict(payload)
 
-        self.count += 1
-        return {"count": self.count}
+            if self.send_async:
+                response = send_payload_async(kwargs={"data": payload})
+            else:
+                response = self.make_request(payload)
+            return {"sent_payload": response}
```
Comment (andrey-canon) [inline webhook_xblock/webhook_xblock.py:245]: if this is sync returns` {"sent_payload": bool}` but if it is async returns {"sent_payload": celerytask}, how do you hanlde that in the front-end

File: webhook_xblock/webhook_xblock.py [hunk 11]
```diff
@@ -47,21 +99,73 @@ def student_view(self, context=None):
...
+                "student": serialized_student.data,
+                "course_id": str(self.runtime.course_id),
+            }
+            payload.update(self.extra_info)
+            payload = flatten_dict(payload)
 
-        self.count += 1
-        return {"count": self.count}
+            if self.send_async:
+                response = send_payload_async(kwargs={"data": payload})
```
Comment (andrey-canon) [inline webhook_xblock/webhook_xblock.py]: this is not async, you have to write something like  func.delay() or func.apply_async()

**Review by andrey-canon — CHANGES_REQUESTED** [review]

File: webhook_xblock/webhook_xblock.py (hunk 9 above)
Comment (magajh) [reply webhook_xblock/webhook_xblock.py]: done

**Comment by magajh** [comment]
I closed this PR accidentally :(
A new PR with these changes has been Opened #2

**Comment by magajh** [comment]
testing comment

File: webhook_xblock/tasks.py (hunk 8 above)
Comment (magajh) [reply webhook_xblock/tasks.py]: Done

**Pull request closed without merging**